proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+- fails due to the arithmetic overflow and throws an exception.
+
+We present two specifications, one for each case.
+
+### transfer success
+
+```{.k .erc20-spec .hkg-spec}
+[transfer-success]
+
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(1, 32) ]
+      _:Map
+    )
+
+```
+
+`<log>` cell specifies that it generates an event log of type `Transfer`.
+
+```{.k .erc20-spec .hkg-spec}
+log:
+      _:List
+    (
+      .List
+    =>
+      ListItem({ ACCT_ID
+               | #parseHexWord(Keccak256(#generateSignature("Transfer", #address(CALLER_ID), #address(TO_ID), #uint256(VALUE))))
+               : CALLER_ID
+               : TO_ID
+               : .WordStack
+               | #asByteStackInWidth(VALUE, 32)
+               })
+    )
+
+refund: _ => _
+
+requires:
+    andBool 0 <=Int TO_ID     andBool TO_ID     <Int (2 ^Int 160)
+    andBool 0 <=Int VALUE     andBool VALUE     <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_FROM  andBool BAL_FROM  <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_TO    andBool BAL_TO    <Int (2 ^Int 256)
+
+comments:
+    // NOTE: negative VALUE is not possible since it is of `num256` type
+```
+
+There are two sub-cases depending on whether the caller is equal to the recipient.
+
+#### transfer success: sub-case 1: the caller `CALLER_ID` is different from the recipient `TO_ID`.
+
+```{.k .erc20-spec .hkg-spec}
+[transfer-success-1]
+```
+
+`<storage>` cell specifies that the amount of `VALUE` is transferred from `balances[CALLER_ID]` to `balances[TO_ID]`.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(balances)}, CALLER_ID) |-> (BAL_FROM => BAL_FROM -Int VALUE)
+```
+
+```{.k .erc20-spec}
+    #hashedLocation({lang}, {pos(balances)}, TO_ID)     |-> (BAL_TO   => BAL_TO   +Int VALUE)
+```
+
+##### HKG-specific:
+On the other hand, HKG token allows the additional overflow, “wrapping around” the balance if it is more than `2^256`, which is very unlikely but still a questionable behavior. See the difference where `+Word`, the addition modulo `2^256`, is used instead of `+Int`, the mathematical, arbitrary precision integer addition.
+
+```{.k .hkg-spec}
+    #hashedLocation({lang}, {pos(balances)}, TO_ID)     |-> (BAL_TO   => BAL_TO  +Word VALUE)
+```
+
+```{.k .erc20-spec .hkg-spec}
+    _:Map
+
++requires:
+    andBool CALLER_ID =/=Int TO_ID
+```
+
+The side-condition ensures that no arithmetic overflow happens.
+
+```{.k .erc20-spec}
+    andBool VALUE <=Int BAL_FROM
+    andBool BAL_TO +Int VALUE <Int (2 ^Int 256) // viper overflow check: (VALUE ==Int 0 xorBool BAL_TO +Word VALUE >Int BAL_TO)
+```
+
+##### HKG-specific:
+As mentioned earlier, HKG token does not check the overflow thus we have no side-condition for that. Instead, HKG token does not allow the zero value to be requested to transfer, which is questionable though, and we have the side-condition for that.
```

Comment:
*the* HKG

Reply:
HKG is not proper ERC20 from what I can see.  I think we should say that upfront and then say that "We can also verify token which are not ERC20-compliant, but we have to change the specification accordingly.  We typically do that together with the customer, to make sure that that' indeed what they want.  It may very well be the case that the customer will actually want to stay ERC20-compliant.  But here we pretend that the HKG customer prefers their non-compliant ERC20 variant."